File: Anurag_19_CSE/Python/day31.py
# Link For Problem: https://leetcode.com/problems/check-if-a-string-contains-all-binary-codes-of-size-k/

class Solution:

    """
        TC : O(nk)

        SC : O(nk)
    """

    def hasAllCodes(self, s: str, k: int) -> bool:
        total = 1 << k

        have = set()

        for i in range(len(s)-k+1):
            temp = s[i:i+k]

            if temp not in have:
                have.add(temp)
                total -= 1

                if total == 0:
                    return True

        return False

File: Anurag_19_CSE/Java/day31.java
//Link For Problem: https://leetcode.com/problems/check-if-a-string-contains-all-binary-codes-of-size-k/

import java.util.HashSet;

public class day31 {

    /*
     * TC : O(nk)
     * 
     * SC : O(nk)
     */

    public boolean hasAllCodes(String s, int k) {
        int need = 1 << k;

        HashSet<String> hs = new HashSet<>();

        for (int i = 0; i < s.length() - k + 1; i++) {
            String temp = s.substring(i, i + k);

            if (!hs.contains(temp)) {
                hs.add(temp);
                need -= 1;

                if (need == 0)
                    return true;
            }
        }

        return false;
    }
}
